_refactor/combined.py

```diff
@@ -0,0 +1,589 @@
+        )
+        rand_y = np.random.uniform(
+            low=np.min(surfsite_pos[:, 1]), high=np.max(surfsite_pos[:, 1]), size=(k,)
+        )
+        rand_z = np.array(
+            [
+                A * sitex + B * sitey + C + added_z
+                for sitex, sitey in zip(rand_x, rand_y)
+            ]
+        )
+        sites = list(np.c_[rand_x, rand_y, rand_z])
+        return sites
+
+    def place_adsorbate_on_sites(self, adsorbate, surface, random_sites):
+        """
+        Place an adsorbates at a list of adsorption sites.
+        Args:
+            adsorbate : An `ase.Atoms` object of the adsorbate
+            surface : An `ase.Atoms` object of the surface
+            random_sites : A list of cartesian coordinates representing the random sites
+
+        Returns:
+            adslabs   A list of `ase.Atoms` object. Each object is an adsorbate placed
+                      on the surface
+        """
+        adslabs = []
+        for site in random_sites:
+            # Add adsorbate to a site
+            adsorbate_c = adsorbate.copy()
+            surface_c = surface.copy()
+            adsorbate_c.translate(site)
+            adslab = surface_c + adsorbate_c
+            tags = [2] * len(adsorbate)
+            final_tags = list(surface.get_tags()) + tags
+
+            # torsion rotate adsorbate if "rotate_adsorbate" is set to True
+            if self.rotate_adsorbate:
+                adsorbate_idx = [idx for idx, tag in enumerate(final_tags) if tag == 2]
+                struct = AseAtomsAdaptor.get_structure(adslab)
+                rotation = np.random.uniform(0, 2 * np.pi)
+                struct.rotate_sites(adsorbate_idx, rotation, (0, 0, 1), site)
+                adslab = AseAtomsAdaptor.get_atoms(struct)
+                adslab.set_tags(final_tags)
+
+            adslab.cell = surface.cell
+            adslab.pbc = [True, True, False]
+            adslabs.append((site, adslab))
+        return adslabs
+
+    def is_config_reasonable(self, adslab):
+        """
+        Function that check whether the adsorbate placement is reasonable.
+        Two criteria are: 1. The adsorbate should be placed on the slab:
+        the fractional coordinates of the adsorption site is bounded by the unit cell.
+        2. The adsorbate should not be buried into the surface: for any atom
+        in the adsorbate, if the distance between the atom and slab atoms
+        are closer than 80% of their expected covalent bond, we reject that placement.
+
+        Args:
+            adslab          An `ase.Atoms` object of the adsorbate+slab complex.
+
+        Returns:
+            A boolean indicating whether or not the adsorbate placement is
+            reasonable.
+        """
+        vnn = VoronoiNN(allow_pathological=True, tol=0.2, cutoff=10)
+        adsorbate_indices = [atom.index for atom in adslab if atom.tag >= 2]
+        adsorbate_bond_indices = [atom.index for atom in adslab if atom.tag == 3]
```

Comment:
where is this set to be 3?

Reply:
`adsorbate_bond_indices` isn't actually used in this method so I think we're okay